---
title: An R Markdown document converted from "dowser_tutorial.ipynb"
output: html_document
---

# Reconstruction and analysis of B-cell lineage trees from single cell data using Immcantation

![](assets/dowser-tutorial-cover.png)

Human B cells play a fundamental role in the adaptive immune response to infection and vaccination, as well as the pathology of allergies and many autoimmune diseases. Central to all of these processes is the fact that B cells are an evolutionary system, and undergo rapid somatic hypermutation and antigen-driven selection as part of the adaptive immune response. The similarities between this B cell response and evolution by natural selection have made phylogenetic methods a powerful means of characterizing important processes, such as immunological memory formation. Recent methodological work has led to the development of phylogenetic methods that adjust for the unique features of B cell evolution. Further, advances in single cell sequencing can now provide an unprecedented resolution of information, including linked heavy and light chain data, as well as the associated transcriptional states of individual B cells. In this tutorial, we show how single cell information can be integrated into B cell phylogenetic analysis using the Immcantation suite (Immcantation.org).

**This tutorial covers:**

Beginning with processed single cell RNA-seq (scRNA-seq) + BCR data from 10X Genomics, we will show:

-   how cell type annotations can be associated with BCR sequences,
-   how clonal clusters can be identified, and
-   how B cell phylogenetic trees can be built and visualized using these data sources.

[Watch on YouTube](https://youtu.be/ZNOSQGh22X4) a recorded version of this tutorial, that was presented at the [Adaptive Immune Receptor Repertoires Webinar Series](https://www.antibodysociety.org/learningcenter/adaptive-immune-receptor-repertoires-webinar-series/) organized by the [AIRR Community](http://airr-community.org/) and [The Antibody Society](https://www.antibodysociety.org/) (November 9, 2021).

## Resources

-   You can email [immcantation\@googlegroups.com](mailto:immcantation@googlegroups.com) with any questions or issues.
-   Documentation: <http://immcantation.org>
-   Source code and bug reports: <https://bitbucket.org/kleinstein/immcantation>
-   Docker/Singularity container for this lab: <https://hub.docker.com/r/immcantation/lab>

## How to use the notebook

Jupyter Notebook documentation: <https://jupyter-notebook.readthedocs.io/en/stable/>

**Ctrl+Enter** will run the code in the selected cell and **Shift+Enter** will run the code and move to the following cell.

## Inside this container

This container comes with software and example data that is ready to use. The commands `versions report` and `builds report` show the versions and dates respectively of the tools and data.

### Software versions

Use this command to list the software versions

```{bash}
versions report
```

### Build versions

Use this command to list the date and changesets used during the image build.

```{bash}
builds report
```

### Example data used in the tutorial

-   `../data/bcr_phylo_tutorial/BCR.data.tsv`: B-Cell Receptor Data. Adaptive Immune Receptor Repertoire (AIRR) tsv BCRs already aligned to IMGT V, D, and J genes. This process is not covered in this tutorial. To learn more visit <https://immcantation.readthedocs.io/en/stable/tutorials/tutorials.html>

-   `../data/bcr_phylo_tutorial/GEX.data.rds`: Gene Expression Data. This file contains a Seurat object with RNA-seq data already processed and annotated. Processing and annotation are not covered in this tutorial. You can learn more on these topics in Seurat's documentation and tutorials: <https://satijalab.org/seurat/articles/pbmc3k_tutorial.html>

These two files are subsamples of the original 10x scRNA-seq and BCR sequencing data from Turner et al. (2020) *Human germinal centres engage memory and naive B cells after influenza vaccination* Nature. 586, 127--132 [link](https://www.nature.com/articles/s41586-020-2711-0) The study consists of blood and limph node samples taken from a single patient at mulitple time points following influenza vaccination.

Note: The example files are available for download [here](https://yale.box.com/shared/static/96y1rlu4ujerprxr53qxw87g7ybc1y99).

## Outline of tutorial

1.  Combining gene expression and BCR sequences.

2.  Identifying clonal clusters, reconstruct germlines.

3.  Building and visualizing trees.

4.  Tree analysis, detecting ongoing evolution.

## The R session

The next two lines of code are required to be able to use R magic and run R code in this Jupyter notebook.

```{python}
# R for Python
# Enable use of R magic to run R code
import rpy2.rinterface
%load_ext rpy2.ipython
```

The is the working directory:

```{r}
getwd()
```

The example files are expected to be in `../data/bcr_phylo_tutorial`.

```{r}
list.files("../data/bcr_phylo_tutorial")
```

## Read in data

The example files are expected to be in `../data/bcr_phylo_tutorial`. If you are using a different path, update the code in the cell bellow accordingly.

```{r}
suppressPackageStartupMessages(library(alakazam))
suppressPackageStartupMessages(library(Seurat))

# Read BCR data
bcr_db <- readChangeoDb("../data/bcr_phylo_tutorial/BCR.data.tsv")

# Read GEX data
gex_db <- readRDS("../data/bcr_phylo_tutorial/GEX.data.rds")
```

## Inspect the data objects

## The gene expression Seurat object

`print` can be used to obtain a general overview of the Seurat object (number of
features, number of samples...).

```{r}
library(Seurat)

# Object summary
print(gex_db)
```

`Idents` reports the cell ID and identities. The first annotation in this blood sample is a TCR.

```{r}
# Cell type annotations
head(Idents(gex_db), 1)
```

## The immune repertoire data

The default file format for all functions in Immcantation is the AIRR-C format as of release 4.0.0. The rearrangement data is stored in a table where each row is a sequence, and each column an annotation fields. To learn more about this format (including the valid field names and their expected values), visit the [AIRR-C Rearrangement Schema documentation](https://docs.airr-community.org/en/stable/datarep/rearrangements.html#fields).

```{r}
suppressPackageStartupMessages(library(dplyr))

# object summary
head(bcr_db, 1)
```

It is possible to subset columns using regular `R` functions. The cell below shows how to subset some fields of interest for the first sequence in the table.

```{r}
# check out select columns
head(select(bcr_db, cell_id, v_call, j_call, sample, day), 1)
```

## Standardize cell IDs

Both of the example datasets have been processed separately, and use slighty different cell identifiers. To consolidate the data into one object, we need to standardize the cell identifiers. This step could be different, or not necessary at all, with other datasets.

```{r}
# Make cell IDs in BCR match those in Seurat Object
bcr_db$cell_id_unique <- paste0(bcr_db$sample, "_", bcr_db$cell_id)
bcr_db$cell_id_unique <- gsub("-1", "", bcr_db$cell_id_unique)
bcr_db$cell_id_unique[1]
```

## Different order

In addition, the cells in both datasets are not presented in the same order.

```{r}
# First id in the BCR data
bcr_db$cell_id_unique[1]
```

```{r}
# First id in the GEX data
Cells(gex_db)[1]
```

Having common cell identifiers, we will be able to bring BCR data into the
Seurat object, or the gene expression and annotation data from the Seurat object
into the BCR table, by matching `cell_id_unique`.

## Add BCR data to Seurat object

1.  Find the GEX cells in the BCR data
2.  Label GEX data with BCR data availability
3.  Plot UMAP

### Find the GEX cells in the BCR data

The vector `match.index` contains the position of the GEX cells in the BCR data. If there is not match, the value will be `NA`.

```{r}
# match index to find the position of the GEX cells in the BCR data
match.index <- match(Cells(gex_db), bcr_db$cell_id_unique)

# In this data, not all cells are B cells. 
# What proportion of cells don’t have BCRs?
mean(is.na(match.index))
```

```{r}
# Just to double check cell ids in the GEX match cells ids in BCR data
# Should be 1
mean(Cells(gex_db) == bcr_db$cell_id_unique[match.index], na.rm = TRUE)
```

### Label GEX data with BCR data availability

With the matching indices, it is possible to label the GEX cells with `TRUE` or `FALSE` to indicate whether there is BCR information available for the cell, and visualize this information in the UMAP plot.

```{r}
# label whether BCR found in cell
gex_db$contains_bcr <- !is.na(match.index)
```

### plot UMAP

We expect that for a large proportion of cells labelled as BCR, there will be BCR sequencing data available, and these cells will be highlighted in the UMAP plot.

```{r}
# List of cells with BCRs
highlighted.cells <- Cells(gex_db)[which(gex_db$contains_bcr)]

# Plot UMAP with BCR-containing cells
DimPlot(object = gex_db, reduction = "umap",
        cells.highlight = highlighted.cells, label = TRUE, cols = "gray",
        pt.size = 1.0, label.size = 8, label.box = TRUE) + NoLegend()
```

## Add GEX data to BCR object

1.  Find the BCR cells in the GEX data
2.  Transfer GEX annotations into the BCR data
3.  Add UMAP coordinates to the BCR data
4.  Remove cells without GEX data
5.  Ensure information transferred from Seurat object

### Find the BCR cells in the GEX data

We repeat the `match` step, reversing the order. The vector `match.index` will now contain the positions of the BCR sequences in the GEX data.

```{r}
# Match indices to find the position of the BCR cells in the GEX data
# Different from finding the position of the GEX cells in the BCR data!
match.index <- match(bcr_db$cell_id_unique, Cells(gex_db))
```

Some BCRs don't have GEX information. This can happen, for example, if the cell for which BCR's are covered didn't pass the GEX processing and quality controls thresholds.

```{r}
# What proportion of BCRs don’t have GEX information?
mean(is.na(match.index))
```

### Transfer GEX annotations into the BCR data

The GEX cell annotations can be added as additional columns in the BCR table.

```{r}
# Add annotations to BCR data
cell.annotation <- as.character(Idents(gex_db))
bcr_db$gex_annotation <- 
   unlist(lapply(match.index,function(x){ifelse(is.na(x), NA,
                                                cell.annotation[x])}))
bcr_db$gex_annotation[1:5]
```

### Add UMAP coordinates to BCR data

The UMAP coordinates can be added as additional columns in the BCR table as well.

```{r}
# Add UMAP coordinates to BCR data
umap1 <- gex_db@reductions$umap@cell.embeddings[, 1]
umap2 <- gex_db@reductions$umap@cell.embeddings[, 2]
bcr_db$gex_umap1 <- unlist(lapply(match.index, function(x){ifelse(is.na(x), NA,
                                                                  umap1[x])}))
bcr_db$gex_umap2 <- unlist(lapply(match.index, function(x){ifelse(is.na(x), NA,
                                                                  umap2[x])}))

bcr_db[1:5, ] %>% select(cell_id_unique, gex_umap1, gex_umap2, gex_annotation)
```

### Remove cells without GEX data

```{r}
# Remove cells that didn’t match
bcr_db = dplyr::filter(bcr_db, !is.na(gex_annotation))
```

### Ensure information transferred from Seurat object

The BCR data table now has the UMAP coordinates. We can reproduce **part** of the UMAP plot with standard `ggplot` commands. This plot will have a similar shape to the GEX UMAP, but will only show points for which both GEX and BCR data is available.

```{r}
suppressPackageStartupMessages(library(ggplot2))

# Set up color palette for annotations
col_anno <- c("GC B" = "dodgerblue2", "PB" = "firebrick2", "ABC" = "seagreen",
              "Naive B" = "darkgoldenrod2", "RMB" = "plum2",
              "Germline" = "black")

# Plot UMAP from bcr_db
bcr_umap <- ggplot(bcr_db) +
               geom_point(aes(x = gex_umap1, y = gex_umap2,
                              color = gex_annotation)) +
               scale_colour_manual(values = col_anno) +
               theme_bw()
bcr_umap
```

### Color the UMAP by isotype

The BCR data frame contains additional annotation fields, such us the isotype,
that can also be visualized on the UMAP. We expect that naive cell express IgM.
The germinal center and plasmablast cells, primarily express IgG and IgA.

```{r}
# Plot isotype on UMAP
ggplot(bcr_db) +
   geom_point(aes(x = gex_umap1,y = gex_umap2, color = isotype)) +
   theme_bw()
```

## Identifying clonal clusters

**Goal:** Partition (cluster) sequences into clonally related lineages. Each
lineage is a group of sequences that came from the same original naive cell.

Summary of the key steps: - *Determine clonal clustering threshold:* sequences
which are under this cut-off are clonally related. - *Assign clonal groups:* add
an annotation (`clone_id`) that can be used to identify a group of sequences
that came from the same original naive cell. - *Reconstruct germline sequences:*
figure out the germline sequence of the common ancestor, before mutations are
introduced during clonal expansion and SMH.

![](assets/clonalexpansion.png)

### Picking a threshold using shazam

*Gupta et al. (2015)*

<img src="assets/hclust.png" width="35%" align="left"/><img src="assets/distNearest.png" width="30%" align="left"/>

We first split sequences into groups that share the same V and J gene
assignments and that have the same junction (or equivalently CDR3) length.
This is based on the assumption that members of a clone will share all of these
properties. `distToNearest` performs this grouping step, then counts the number
of mismatches in the junction region between all pairs of sequences in each
group and returns the smallest non-zero value for each sequence. At the end of
this step, a new column (`dist_nearest`) which contains the distances to the
closest non-identical sequence in each group will be added to the BCR table.
`findThreshold` uses the distribution of distances calculated in the previous
step to determine an appropriate threshold for the dataset. This can be done
using either a `density` or `mixture` based method.

### distToNearest

```{r}
suppressPackageStartupMessages(library(shazam))

# Find threshold using heavy chains
dist_ham <- distToNearest(dplyr::filter(bcr_db, locus == "IGH"))
head(dist_ham) %>% select(cell_id_unique, dist_nearest)
```

### findThreshold

The figure shows the distance-to-nearest distribution for the repertoire. 
Typically, the distribution is bimodal. The first mode (on the left) represents
sequences that have at least one clonal relative in the dataset, while the
second mode (on the right) is representative of the sequences that do not have
any clonal relatives in the data (sometimes called "singletons"). A reasonable
threshold will separate these two modes of the distribution.

```{r}
output <- findThreshold(dist_ham$dist_nearest)
threshold <- output@threshold

# Visualize the distance-to-nearest distribution and threshold
plotDensityThreshold(output)
```

### Performing clustering using scoper

Once a threshold is decided, we perform the clonal assignment. At the end of
this step, the BCR table will have an additional column (`clone_id`) that
provides an identifier for each sequence to indicate which clone it belongs to
(i.e., sequences that have the same identifier are clonally-related). *Note that these identifiers are only unique to the dataset used to carry out the clonal assignments.*

Note: will print out "running in bulk mode" because the subsampled example data
has only heavy chains. [Other options](https://scoper.readthedocs.io/en/stable/topics/hierarchicalClones/#single-cell-data) available if light chains are included.

```{r}
suppressPackageStartupMessages(library(scoper))

# Assign clonal clusters
results <- hierarchicalClones(dist_ham, threshold = threshold)
results_db <- as.data.frame(results)

head(results_db) %>% select(cell_id_unique, clone_id)
```

### Visualize clone size distribution

Most real datasets, will have most clones of size 1 (one sequence). In this
tutorial, we processed data to remove most of singleton clone and we don't see
the much higher peak at 1 that we would normally expect.

```{r}
# get clone sizes using dplyr functions
clone_sizes <- countClones(results_db)

# Plot cells per clone
ggplot(clone_sizes, aes(x = seq_count)) +
   geom_bar() + theme_bw() +
   xlab("Sequences per clone")
```

## Reconstruct clonal germlines using dowser

The goal is to reconstruct the sequence of the unmutated ancestor of each clone.
We use a reference database of known alleles ([IMGT](http://www.imgt.org)).
Because it is very difficult to ccurately infer the D region and the junction
region for BCR sequences, we mask this region with `N`.

Note: If you opted for a native installation to run this tutorial, you can
obtain reference germlines from IMGT with:

        git clone https://bitbucket.org/kleinstein/immcantation
        immcantation/scripts/fetch_imgtdb.sh

```{r}
suppressPackageStartupMessages(library(dowser))

# read in IMGT data if downloaded on your own (above)
# and update `dir` to use the path to your `human/vdj` folder
# references = readIMGT(dir = "human/vdj/")

# Read in IMGT data if using in Docker image
references = readIMGT(dir = "/usr/local/share/germlines/imgt/human/vdj")

# Reconstruct germlines
results_db = createGermlines(results_db, references)

# Check output column
results_db$germline_alignment_d_mask[1]
```

# Building and visualizing trees

1.  Formatting clones
2.  Tree building
3.  Visualize trees
4.  Reconstruct intermediate sequences

## Formatting clones with dowser

In the rearrangement table, each row corresponds to a sequence, and each column
is information about that sequence. We will create a new data structure, where
each row is a clonal cluster, and each column is information about that clonal
cluster. The function `formatClones` performs this processing and has options
that are relevant to determine how the trees can be built and visualized. For
example, `traits` determines the columns from the rearrangement data that will
be included in the `clones` object, and will also be used to determine the
uniqueness of the sequences, so they are not collapsed.

```{r}
# Make clone objects with aligned, processed sequences
# collapse identical sequences unless differ by trait
# add up duplicate_count column for collapsed sequences
# store day, isotype, gex_annotation
# discard clones with < 5 distinct sequences
clones <- formatClones(results_db,
                       traits = c("day", "isotype", "gex_annotation"),
                       num_fields = c("duplicate_count"), minseq = 5)
clones
```

## Tree building with dowser

Dowser offers multiple ways to build B cell phylogenetic trees. These differ by the method used to estimate tree topology and branch lengths (e.g. maximum parsimony and maximum likelihood) and implementation (IgPhyML, PHYLIP, or R packages ape and phangorn). Each method has pros and cons.

### Maximum parsimony

This is the oldest method and very popular. It tries to minimize the number of mutations from the germline to each of the tips. It can produce misleading results when parallel mutations are present.

```{r}
# Two options for maximum parsimony trees
# 1. phangorn
trees <- getTrees(clones)
```

```{r}
# 2. dnapars (PHYLIP)
trees <- getTrees(clones, build = "dnapars", exec = "/usr/local/bin/dnapars")
head(trees)
```

### Standard maximum likelihood

These methods model each sequence separately. Use a markov model of the mutation process and try to find the tree, not the branch lengths, that maximizes the likehood of seen data.

```{r}
# Two options for standard maximum likelihood trees
# 1. pml (phangorn)
trees <- getTrees(clones, build = "pml", sub_model = "GTR")
head(trees)
```

```{r}
# 2. dnaml (PHYLIP)
trees <- getTrees(clones, build = "dnaml", exec = "/usr/local/bin/dnaml")
head(trees)
```

### Bcell specific maximum likelihood

Similar to the standard maximum likelihood, but incorporating SHM specific mutation biases into the tree building.

```{r}
# B cell specific maximum likelihood with IgPhyML
# This code is commented out because of long running time
# trees = getTrees(clones, build="igphyml", exec="/usr/local/share/igphyml/src/igphyml", nproc=2)
# head(trees)
```

## Plotting trees with dowser and ggtree

All tree building methods are plotted using the same method in dowser, `plotTrees`.

```{r}
# Plot all trees
plots = plotTrees(trees, tips = "isotype", tipsize = 2)
```

```{r}
# Plot the largest tree
plots[[1]]
```

```{r}
# Save PDF of all trees
dir.create("results/dowser_tutorial/", recursive = TRUE)
treesToPDF(plots,
           file = file.path("results/dowser_tutorial/final_data_trees.pdf"),
           nrow = 2, ncol = 2)
```

## More elaborate tree plots

Plot trees so that tips are colored by cell type, scaled by sample day, and labelled by isotype.

```{r}
suppressPackageStartupMessages(library(ggtree))

# Scale branches to mutations rather than mutations/site
trees = scaleBranches(trees)

# Make fancy tree plot of second largest tree
plotTrees(trees, scale = 5)[[2]] +
   geom_tippoint(aes(colour = gex_annotation, size = day)) +
   geom_tiplab(aes(label = isotype), offset = 0.002) +
   scale_colour_manual(values = col_anno)
```

## Reconstruct intermediate sequences

Get the predicted intermediate sequence at an internal node in the second largest tree. Dots represent IMGT gaps.

```{r}
suppressPackageStartupMessages(library(ggtree))

# Collapse nodes with identical sequences
trees = collapseNodes(trees)

# node_nums=TRUE labels each internal node
p <- plotTrees(trees, node_nums = TRUE, labelsize = 6, scale = 5)[[2]] +
       geom_tippoint(aes(colour = gex_annotation, size = day)) + 
       geom_tiplab(aes(label = isotype), offset = 0.002) + 
       scale_colour_manual(values = col_anno)
print(p)

# Get sequence at node 26 for the second clone_id in trees
getSeq(trees, clone = trees$clone_id[2], node = 26)
```

## Test for measurable evolution

Perform root-to-tip regression on each tree to detect if later-sampled timepoints are more diverged from the germline. See this reference for more detail:

Hoehn, K. B. et al. (2021) Human B cell lineages engaged by germinal centers following influenza vaccination are measurably evolving. bioRxiv. <https://doi.org/10.1101/2021.01.06.425648>

```{r}
# Correlation test
trees = correlationTest(trees, time = "day")

# Remove trees with one timepoint, order by p value
trees = dplyr::filter(trees, !is.na(p))
trees = trees[order(trees$p), ]

# Fancy tree plots coloring tips by sample day
p = plotTrees(trees)
p = lapply(p, function(x)
   x + 
      geom_tippoint(aes(fill = day), shape = 21, size = 3) +
      scale_fill_distiller(palette = "RdYlBu"))

# Save all trees to a pdf file
treesToPDF(p, file = "results/dowser_tutorial/time_data_trees.pdf")

select(trees, clone_id, slope, correlation, p)
print(p[[1]])
```

# References

## B cell phylo

Hoehn, K. B. et al. (2016) The diversity and molecular evolution of B-cell receptors during infection. MBE. <https://doi.org/10.1093/molbev/msw015>

Hoehn, K. B. et al. (2019) Repertoire-wide phylogenetic models of B cell molecular evolution reveal evolutionary signatures of aging and vaccination. PNAS 201906020.

Hoehn, K. B. et al. (2020) Phylogenetic analysis of migration, differentiation, and class switching in B cells. bioRxiv. <https://doi.org/10.1101/2020.05.30.124446>

Hoehn, K. B. et al. (2021) Human B cell lineages engaged by germinal centers following influenza vaccination are measurably evolving. bioRxiv. <https://doi.org/10.1101/2021.01.06.425648>

## BCR analysis

Gupta,N.T. et al. (2017) Hierarchical clustering can identify b cell clones with high confidence in ig repertoire sequencing data. The Journal of Immunology, 1601850.

Gupta,N.T. et al. (2015) Change-o: A toolkit for analyzing large-scale b cell immunoglobulin repertoire sequencing data. Bioinformatics, 31, 3356--3358.

Nouri,N. and Kleinstein,S.H. (2018a) A spectral clustering-based method for identifying clones from high-throughput b cell repertoire sequencing data. Bioinformatics, 34, i341--i349.

Nouri,N. and Kleinstein,S.H. (2018b) Optimized threshold inference for partitioning of clones from high-throughput b cell repertoire sequencing data. Frontiers in immunology, 9.

Stern,J.N. et al. (2014) B cells populating the multiple sclerosis brain mature in the draining cervical lymph nodes. Science translational medicine, 6, 248ra107--248ra107.

Vander Heiden,J.A. et al. (2017) Dysregulation of b cell repertoire formation in myasthenia gravis patients revealed through deep sequencing. The Journal of Immunology, 1601415.

Yaari,G. et al. (2012) Quantifying selection in high-throughput immunoglobulin sequencing data sets. Nucleic acids research, 40, e134--e134.

Yaari,G. et al. (2013) Models of somatic hypermutation targeting and substitution based on synonymous mutations from high-throughput immunoglobulin sequencing data. Frontiers in immunology, 4, 358.
